executor/replace.go

```diff
@@ -57,46 +108,67 @@ func (e *ReplaceExec) exec(rows []types.DatumRow) error {
 	 * because in this case, one row was inserted after the duplicate was deleted.
 	 * See http://dev.mysql.com/doc/refman/5.7/en/mysql-affected-rows.html
 	 */
-	idx := 0
-	rowsLen := len(rows)
-	sc := e.ctx.GetSessionVars().StmtCtx
-	for {
-		if idx >= rowsLen {
+	err := e.batchGetInsertKeys(e.ctx, e.Table, newRows)
+	if err != nil {
+		return errors.Trace(err)
+	}
+
+	// Batch get the to-be-replaced rows in storage.
+	err = e.initDupOldRowValue(e.ctx, e.Table, newRows)
+	if err != nil {
+		return errors.Trace(err)
+	}
+
+	for i, r := range e.toBeCheckedRows {
+		for {
```

Comment:
Can we extract the inner for loop into its own function.
It's hard to see where the code is going to execute after `break` and `continute`.